File: garuda/milestone1-ingestion/02-alertmanager-bridge/internal/schema/schema.go
// Package schema defines the canonical AnomalySignal envelope and the
// fingerprint helper used by the Alertmanager bridge (Tier 1 producer).
//
// The schema mirrors _shared/schemas/anomaly_signal.schema.json (v1).
// Keep the two in sync; the Tier 2 Python worker has the equivalent
// Pydantic model in 03-metrics-collector/shared/schema.py.
package schema

import (
	"crypto/sha256"
	"encoding/hex"
	"sort"
	"strings"
	"time"
)

// SchemaVersion is the wire-protocol version of the AnomalySignal envelope.
// Bump only on backward-incompatible changes.
const SchemaVersion = "v1"

// AnomalySignal is the canonical envelope written to garuda.anomalies.tier1.
// JSON tag order matches the JSON Schema for human-readable output.
type AnomalySignal struct {
	SchemaVersion string            `json:"schema_version"`
	AnomalyID     string            `json:"anomaly_id"`
	Tier          int               `json:"tier"`
	Detector      string            `json:"detector"`
	TsDetected    time.Time         `json:"ts_detected"`
	TsStarted     time.Time         `json:"ts_started"`
	TsResolved    *time.Time        `json:"ts_resolved,omitempty"`
	Fingerprint   string            `json:"fingerprint"`
	Metric        Metric            `json:"metric"`
	Cluster       string            `json:"cluster"`
	Namespace     *string           `json:"namespace,omitempty"`
	Service       *string           `json:"service,omitempty"`
	Severity      string            `json:"severity"`
	Confidence    float64           `json:"confidence"`
	Values        map[string]any    `json:"values"`
	Evidence      Evidence          `json:"evidence"`
	SourceLabels  map[string]string `json:"source_labels,omitempty"`
	Annotations   map[string]string `json:"annotations,omitempty"`
}

// Metric describes the metric the anomaly relates to.
type Metric struct {
	Name     string            `json:"name"`
	Matchers map[string]string `json:"matchers"`
}

// Evidence holds the supporting context for the anomaly.
type Evidence struct {
	PromQL  *string  `json:"promql,omitempty"`
	Window  Window   `json:"window"`
	Samples [][2]any `json:"samples,omitempty"` // (timestamp string, value number)
}

// Window is the time window used to evaluate / observe the anomaly.
type Window struct {
	Start time.Time `json:"start"`
	End   time.Time `json:"end"`
}

// Fingerprint computes the deterministic 16-char hex hash documented in
// _shared/schemas/README.md.
//
//	input  = detector + "|" + metric_name + "|" + sorted_kv(matchers) + "|" + cluster
//	hash   = sha256(input)
//	output = first 16 hex chars
//
// Producers MUST use this exact algorithm so Tier 1 and Tier 2 detections
// of the same underlying issue collide on fingerprint and the correlator
// can fuse them.
func Fingerprint(detector, metricName string, matchers map[string]string, cluster string) string {
	keys := make([]string, 0, len(matchers))
	for k := range matchers {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var sb strings.Builder
	for i, k := range keys {
		if i > 0 {
			sb.WriteByte(',')
		}
		sb.WriteString(k)
		sb.WriteByte('=')
		sb.WriteString(matchers[k])
	}

	input := detector + "|" + metricName + "|" + sb.String() + "|" + cluster
	sum := sha256.Sum256([]byte(input))
	return hex.EncodeToString(sum[:])[:16]
}

File: garuda/milestone1-ingestion/03-metrics-collector/garuda_tier2/schema.py
"""AnomalySignal envelope (Tier 2 producer).

Mirrors:
- internal/schema/schema.go (Go, Tier 1 bridge)
- _shared/schemas/anomaly_signal.schema.json (canonical JSON Schema)

The fingerprint algorithm here MUST match the Go implementation byte-for-byte
so a Tier 1 alert and a Tier 2 ML detection on the same underlying anomaly
collide on fingerprint and the correlator can fuse them.
"""

from __future__ import annotations

import hashlib
import uuid
from datetime import datetime, timezone
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

SCHEMA_VERSION = "v1"


def fingerprint(detector: str, metric_name: str, matchers: dict[str, str], cluster: str) -> str:
    """Deterministic 16-char hex hash. Same algo as Go's schema.Fingerprint.

    >>> fingerprint("d", "m", {"k": "v"}, "c") == fingerprint("d", "m", {"k": "v"}, "c")
    True
    """
    keys = sorted(matchers)
    parts = ",".join(f"{k}={matchers[k]}" for k in keys)
    inp = f"{detector}|{metric_name}|{parts}|{cluster}"
    return hashlib.sha256(inp.encode("utf-8")).hexdigest()[:16]


def now_utc() -> datetime:
    return datetime.now(tz=timezone.utc)


def new_anomaly_id() -> str:
    return str(uuid.uuid4())


class Window(BaseModel):
    model_config = ConfigDict(extra="forbid")
    start: datetime
    end: datetime


class Evidence(BaseModel):
    model_config = ConfigDict(extra="forbid")
    promql: str | None = None
    window: Window
    samples: list[tuple[datetime, float]] | None = None


class Metric(BaseModel):
    model_config = ConfigDict(extra="forbid")
    name: str
    matchers: dict[str, str]


class AnomalySignal(BaseModel):
    """Canonical envelope. Field names + JSON shape match the schema file."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: str = SCHEMA_VERSION
    anomaly_id: str = Field(default_factory=new_anomaly_id)
    tier: int  # always 2 for this producer; left explicit for clarity
    detector: str
    ts_detected: datetime
    ts_started: datetime
    ts_resolved: datetime | None = None
    fingerprint: str
    metric: Metric
    cluster: str
    namespace: str | None = None
    service: str | None = None
    severity: str
    confidence: float = Field(ge=0.0, le=1.0)
    values: dict[str, Any] = Field(default_factory=dict)
    evidence: Evidence
    source_labels: dict[str, str] = Field(default_factory=dict)
    annotations: dict[str, str] = Field(default_factory=dict)

    def to_kafka_bytes(self) -> bytes:
        return self.model_dump_json(exclude_none=True).encode("utf-8")
